### ISSUE TITLE
`ts/quote-props`: Extend eslint core's `quote-props` rule

### ISSUE THREAD

**Issue by so1ve** [eslint-stylistic/eslint-stylistic#274]
### Clear and concise description of the problem

```json
{
  "rules": {
    "quote-props": ["error", "as-needed"],
  }
}
```

```ts
const obj1 = {a: 42}; // OK
const obj2 = {"b": 42}; // ESLint: Unnecessarily quoted property 'b' found.(quote-props)
type obj1 = {a: number}; // OK
type obj2 = {"b": number}; // Expected ESLint error here
enum Foo { a = 1 } // OK
enum Foo { "a" = 1 } // Expected ESLint error here
```

### Suggested solution

Implement https://github.com/typescript-eslint/typescript-eslint/issues/3301

### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Contributes

- [X] If this feature request is accepted, I am willing to submit a PR to fix this issue

**Comment by antfu** [comment]
Sounds good, PR welcome :)
